MaxText/layers/pipeline.py

```diff
@@ -210,32 +231,38 @@ def get_new_loop_state(self,output, loop_state):
       * state_io: rotates left/up by 1 (the whole created in the last slot is filled with the most recent pipeline output)
          * Pushing inputs up from top of state_io into first stage of shift
          * Pulling outputs up from last stage of shift into bottom of state_io
-      * shift: rotate output right/down by 1 - we imagine the pipeline moves to right/down
+      * shift: rotate output (or prev_outputs if using delay) right/down by 1 - we imagine the pipeline moves to right/down
       * circ_storage: pushes circ_storage_mover (the output of the previous iteration) into rotating index of circ_storage
       * circ_storage_mover: assigned to rotated output and pushed into circ_storage on the next iteration
+      * prev_outputs: is set to the current output
     '''
 
     old_state_io = loop_state['state_io']
     old_circ_storage = loop_state["circ_storage"]
     old_circ_storage_mover = loop_state["circ_storage_mover"]
     loop_iteration = loop_state["loop_iteration"]
+    old_prev_outputs = loop_state["prev_outputs"]
     # Shift becomes a rotated-right version of the previous output
     def _rotate_right(output_in):
       # Use lax.slice to avoid generating a gather.
       last = jax.lax.slice_in_dim(output_in, self.num_stages - 1, self.num_stages, axis=0)
       except_last = jax.lax.slice_in_dim(output_in, 0, self.num_stages - 1, axis=0)
       return jnp.concatenate([last, except_last], axis=0)
-    new_shift = _rotate_right(output)
+    if self.config.pipeline_delay_activation_forwarding:
+      new_shift = _rotate_right(old_prev_outputs)
```

Comment:
Might be missing something, but where do the inputs get rotated into `prev_outputs`?

Reply:
inputs -> outputs -> new_prev_outputs

Reply:
Synced offline - we're using `prev_outputs` to delay the next stage's compute on this iteration's output, so `prev_outputs` is just an extra intermediary between `outputs` and `shift`.